types/underscore/underscore-tests.ts

```diff
@@ -659,3 +659,129 @@ function strong_typed_values_tests() {
 
     _.values<{title: string, value: number}>(dictionaryLike);
 }
+
+// test for #7931 - verify that the result of a function like reduce that returns a singleton can be chained further
+// $ExpectType number[]
+_.chain([1, 2, 3])
+    .reduce((acc, x) => [x], [] as number[])
+    .map(() => 1)
+    .value();
```

Comment:
The problem with this example from #7931, though, is that while it passes with the current changes it's not a great example of what someone would actually want to do since most folks would want to use the values from the reduce call rather than just turning them all into ones. I tried doing something that someone would actually want to do in such a situation (shown below) and ran into trouble. Happily, it's trouble that I know how to deal with pretty quickly and easily from my long-term experimentation! Less happily, applying it will require a change to all return types that were previously `_ChainSingle`s and potentially the addition of more tests 😞.

```TS
_.chain([1, 2, 3])
    .reduce((acc, x) => { acc.unshift(x); return acc; }, [] as number[])
    .map(n => n + 1) // error - n is number[] when it should be number
    .value();
```

Since I know what to do here, it won't be difficult to do, and it would be nice for this to work in situations like those described in the issue filed around this, I'll make that change and add a few more interesting tests around that. Sorry for not catching this until now 😞

Reply:
I should note, too, that this is specifically an issue with operations that result in collections; these changes would still work fine as-is with reduce and other operations previously resulting in a `_ChainSingle` that would yield non-collection items like objects with properties of multiple types, strings, and numbers.

Reply:
The change is made! To do this, I reverted the change where I removed `_ChainSingle<T>` and instead redefined `_ChainSingle<V>` as `_Chain<TypeOfCollection<V>, V>`, which is admittedly how I ideally see `_Chain` itself being defined long-term anyway except with `V` being the first parameter and `T` technically not even needing to exist. One other advantage around making this change is that it reduces the likelihood of accidental foot-shooting because any type `V` that is not a collection will result in a `T` type of `never`, which makes calling the functions that expect to act on collections not work very well when tried on wrapped non-collections 😁 

This also involved updating `Collection<T>` to something more useful, though I was planning on doing that in the next PR anyway.